src/main/java/seedu/address/logic/commands/ChangeCommand.java

```diff
@@ -0,0 +1,208 @@
+            + "by the index number used in the displayed appointment list. "
+            + "A new appointment will be created with the input values.\n"
+            + "The old appointment at the input index will be deleted from the appointment book.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_DATE + "DATE] "
+            + "[" + PREFIX_TIME + "TIME] "
+            + "[" + PREFIX_DURATION + "DURATION] \n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_DATE + "12-Dec-2021 "
+            + PREFIX_TIME + "4:00PM"
+            + PREFIX_DURATION + "30";
+
+    public static final String MESSAGE_EDIT_APPOINTMENT_SUCCESS = "Edited Appointment: %1$s";
+    public static final String MESSAGE_DUPLICATE_APPOINTMENT = "This appointment already exists in the appointment "
+            + "book.";
+
+    public static final String ASSIGNMENT_OVERLAP = "This time slot is occupied";
+    public static final String DATE_MISSING = "The date of appointment is missing";
+    public static final String TIME_MISSING = "The time of appointment is missing";
+    public static final String DURATION_MISSING = "The duration of appointment is missing";
+
+    private final Index index;
+    private final EditAppointmentDescriptor editAppointmentDescriptor;
+
+    /**
+     * @param index of the appointment in the filtered appointment list to edit
+     * @param editAppointmentDescriptor details to edit the appointment with
+     */
+    public ChangeCommand(Index index, EditAppointmentDescriptor editAppointmentDescriptor) {
+        requireNonNull(index);
+        requireNonNull(editAppointmentDescriptor);
+
+        this.index = index;
+        this.editAppointmentDescriptor = new EditAppointmentDescriptor(editAppointmentDescriptor);
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        List<Appointment> lastShownList = model.getFilteredAppointmentList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_APPOINTMENT_DISPLAYED_INDEX);
+        }
+
+        Appointment appointmentToEdit = lastShownList.get(index.getZeroBased());
+        Appointment editedAppointment = createEditedAppointment(appointmentToEdit, editAppointmentDescriptor);
+
+        if (appointmentToEdit.equals(editedAppointment) && model.hasAppointment(editedAppointment)) {
+            throw new CommandException(MESSAGE_DUPLICATE_APPOINTMENT);
+        }
+        model.deleteAppointment(appointmentToEdit);
+        if (model.hasOverlappingAppointment(editedAppointment)) {
+            model.addAppointment(appointmentToEdit);
+            throw new CommandException(ASSIGNMENT_OVERLAP);
+        }
+        model.addAppointment(appointmentToEdit);
+        model.setAppointment(appointmentToEdit, editedAppointment);
```

Comment:
Remember to edit this part 👍  
The logic for setting appointments can be handled in `UniqueAppointmentList` and this part can be edited to catching `OverlappingAppointmentException` and throwing the relevant `CommandException`.